### PR TITLE
Plugin based setup assistant 

### PR DESCRIPTION
**Pull request by MohmadAyman**
This PR attempts to implement all MSA widgets as plugins. It retargets #1243 at melodic-devel.

- Setup screen widget class is the base for the plugins.
- Each widget is turned to a plugin and is now in `moveit_setup_assistant_plugins`
- Plugins are loaded in the `SetupAssistantWidget`.
- Numbers are added to the widgets names to enforce the order in the navigation menu.

### CODE REVIEW

**Comment by MohmadAyman** [comment]
I really can't count the number of times I had to resolve merge conflicts to get this PR ready, it has been three months, can you please please review and close it??

**Comment by MohmadAyman** [comment]
@rhaschke @davetcoleman  @gavanderhoorn please review.
@davetcoleman I've addressed your comments from pr #1243 .

**Comment by MohmadAyman** [comment]
Travis says
```
moveit_setup_assistant: CMakeLists.txt: error: moveit_setup_assistant plugin file 'plugins/author_information_widget/author_information_widget.xml' is not installed to ${CATKIN_PACKAGE_SHARE_DESTINATION}
```
Is that a wanted behavior? 
I do install the `${CATKIN_PACKAGE_SHARE_DESTINATION}` in `plugins/author_information_widget/CMakeLists.txt`, which looks cleaner since we have 11 widgets.
Does this install have to be in the higher level `CmakeLists.txt`?

**Comment by rhaschke** [comment]
@MohmadAyman, could you please report the issue at https://github.com/fkie/catkin_lint.

**Comment by rhaschke** [comment]
> I do install the ${CATKIN_PACKAGE_SHARE_DESTINATION} in plugins/author_information_widget/CMakeLists.txt, which looks cleaner since we have 11 widgets.
Does this install have to be in the higher level CmakeLists.txt?

@MohmadAyman, I think these .xml files need to reside at the top-level to be properly found in devel-space mode as well. That's why `catkin_lint` might only check for their installation from the top-level `CMakeLists.txt`. My suggestion is to reduce to a single .xml file listing all plugins as, for example, rviz does too.

**Comment by MohmadAyman** [comment]
> As far as I can see, your approach doesn't consider data modularization yet: All the config data is stored in one central location, the MoveItConfigData structure, and all files are created in one central widget, the ConfigurationFilesWidget.
For this reason, it's not yet possible to add a self-contained third-party widget plugin creating its own set of config files, which I think is the key motivation for introducing widget plugins. But, maybe, you have another motivation?

While I do agree this PR doesn't make the SA as modular as we would want it to be, but I would propose to do the API change for the config data and the writing of the configuration files in another PR. I think this PR is already too big. 
So I propose this PR would just be a first step by turning the existing widgets into plugins, while everything else stays the same, and then a follow up PR focused on changing the API to make it modular for usage. @rhaschke

**Comment by rhaschke** [comment]
I like to keep related changes together, so I don't support the idea of accepting this for the master branch now and have one or more related commits in the future. But I also see the point of large changesets, which increase the reviewing burden. That's why I suggest the following: We merge this PR (squashed into a single commit) into a new feature branch, and you can later file additional PRs to this branch. Finally, if the development of the modular MSA branch is finished, we merge the whole branch into master.

**Comment by MohmadAyman** [comment]
Merging the PR into a feature branch does sound good.
I will check your comments on the code and get back to you.

**Comment by davetcoleman** [comment]
@rhaschke thank you for your review and suggestions!

> So I propose this PR would just be a first step by turning the existing widgets into plugins, while everything else stays the same, and then a follow up PR focused on changing the API to make it modular for usage. 

Adding this to a feature branch won't help with the issue of it continuing to have merge conflicts, especially with the clang-tidy efforts going on.

Assuming the current approach doesn't have any bugs / feature regressions, I see no reason we can't merge this now and improve the way plugins are self-contained later. I strongly agree with @MohmadAyman that the PR should remain at its current size / manageable. Asking for more and more improvements of volunteer maintainers is a good way to scare them off.

**Comment by rhaschke** [comment]
> Adding this to a feature branch won't help with the issue of it continuing to have merge conflicts.

That's true, but I volunteer to resolve those conflicts later on.

**Comment by rhaschke** [comment]
@MohmadAyman, I created the branch `feature-modular-msa` and retargeted this PR against this new branch.

**Comment by MohmadAyman** [comment]
Thank you, I will review the code comments and ping you when I'm done addressing them.

File: moveit_setup_assistant/plugins/planning_groups_widget/planning_groups_widget.cpp [hunk 1]
```diff
@@ -129,7 +138,7 @@ PlanningGroupsWidget::PlanningGroupsWidget(QWidget* parent, const MoveItConfigDa
           SLOT(previewSelectedLink(std::vector<std::string>)));
 
   // Chain Widget
-  chain_widget_ = new KinematicChainWidget(this, config_data);
+  chain_widget_ = new KinematicChainWidget(this, config_data_);
```
[the remark this answers is not in the document]
Comment (MohmadAyman) [reply moveit_setup_assistant/plugins/planning_groups_widget/planning_groups_widget.cpp]: It is initialized in `initializeWidget()` which is the first function to get called.

File: moveit_setup_assistant/src/widgets/setup_assistant_widget.cpp [hunk 2]
```diff
@@ -242,112 +231,60 @@ void SetupAssistantWidget::progressPastStartScreen()
...
-  connect(author_information_widget_, SIGNAL(highlightLink(const std::string&, const QColor&)), this,
-          SLOT(highlightLink(const std::string&, const QColor&)));
-  connect(author_information_widget_, SIGNAL(highlightGroup(const std::string&)), this,
-          SLOT(highlightGroup(const std::string&)));
-  connect(author_information_widget_, SIGNAL(unhighlightAll()), this, SLOT(unhighlightAll()));
-
-  // Configuration Files
-  configuration_files_widget_ = new ConfigurationFilesWidget(this, config_data_);
-  main_content_->addWidget(configuration_files_widget_);
+  if (!plugins_loader_)
```
[the remark this answers is not in the document]
Comment (MohmadAyman) [reply moveit_setup_assistant/src/widgets/setup_assistant_widget.cpp]: Actually yes it seems to be only entered once, I'll remove the guard.

File: moveit_setup_assistant/package.xml [hunk 3]
```diff
@@ -31,9 +31,26 @@
...
+  <!-- Optional: Gazebo integration. This is a heavy dependency so is disabled by default -->
+  <!-- depend>gazebo_ros</depend -->
 
   <exec_depend>xacro</exec_depend>
 
   <test_depend>moveit_resources</test_depend>
   <test_depend>rosunit</test_depend>
+
+  <export>
+    <moveit_setup_assistant plugin="${prefix}/plugins/author_information_widget/author_information_widget.xml"/>
```
[the remark this answers is not in the document]
Comment (MohmadAyman) [reply moveit_setup_assistant/package.xml]: is there any particular reason to do that?. the structure looks clean with each directory contain all the plugin related stuff including the `.xml`.

[the remark this answers is not in the document]
Comment (rhaschke) [reply moveit_setup_assistant/package.xml]: Please see https://github.com/ros-planning/moveit/pull/1375#issuecomment-471315162.

**Comment by MohmadAyman** [comment]
So I've drafted an approach to implement this, it doesn't involve loading data to the new (non default) plugins (widgets) yet, as it is not even the case for default widgets at the current stage of the SA,

In moveit_config_data
```
[New variable] some_data_structure extra_configs_;
[New function] registerConfig(YAML::Emitter& emitter, std::string config_name) {adds the emitter to the extra_config_ data structure}
```

configuration_files_widget
```
checks the extra_configs if it is not empty, and process the files there.
```

for a new plugin
```
has to construct their config file, then call moveit_config_data->registerConfig(YAML::Emitter& emitter, std::string config_name), passing the file to it
```

does this sound like a good approach? any suggestion?

**Comment by rhaschke** [comment]
> So I've drafted an approach to implement this ...

@MohmadAyman, what exactly do you mean by "this"? Please refer to existing comments or use quoting to reference previous posts.
Please focus for now on getting this PR merged into the `feature-modular-msa` branch. From there, you can continue to further modularize.

**Comment by MohmadAyman** [comment]
> Hence, you should also propose a mechanism to modularize config file creation and data storage.
> I suggest to have additional virtual functions, loadConfig and saveConfig, loading existing old configuration files (from a previous run of MSA) and saving new files. If loadConfig doesn't consider all manual changes applied to these files (as it is currently the case), a corresponding warning needs to be issued in the ConfigurationFilesWidget. To this end, the ConfigurationFilesWidget needs to be extended to provide an API to "register" config files to be created. There should also be a mechanism to protect multiple widget plugins writing the same file, thus overriding each other.
> 
> I think, there should be some general data available to all widgets, like the RobotModel. Most other data should be local to the widgets. What do you think?

[my reply](https://github.com/ros-planning/moveit/pull/1375#issuecomment-478287559)

**Comment by rhaschke** [comment]
Thanks for clarification.

**Comment by MohmadAyman** [comment]
> Thanks for clarification.

Hey @rhaschke would you please clarify, does this mean you agree with this approach or against it?

**Comment by rhaschke** [comment]
Dear @MohmadAyman,
sorry, I was looking for a new commit and didn't recognized [your reply](https://github.com/ros-planning/moveit/pull/1375#issuecomment-478287559) as _the proposal_.

Until recently, all files generated by MSA were obligatory. However, with your recent additions, some files, e.g. simulation, perception, or controller config files, are optional, right?
Those files should be only generated (and thus listed in the ConfigurationFilesWidget) if the corresponding settings are actually configured. Alternatively, they could be disabled (greyed out) there.

> [New variable] some_data_structure extra_configs_;
> [New function] registerConfig(YAML::Emitter& emitter, std::string config_name)

Registering a YAML emitter seems to be too restrictive here - we do have XMLs and maybe other file types as well. I suggest to register the (absolute) file name only. 
The `ConfigurationFilesWidget` should call `SetupScreenWidget::saveConfig()` for each file.

However, I propose another procedure to progress:
1. Finish this PR. There are some general remarks in https://github.com/ros-planning/moveit/pull/1375#pullrequestreview-212596707 to address before we can merge this into the created feature branch.
2. Add loading of existing config files to allow proper editing of files. This will render the ConfigurationFilesWidget obsolete as there is no danger to overwrite manual changes.
3. Optionally, think about using Qt's UI files to layout widgets - instead of manually creating and layouting items.
4. Fully modularize MSA widgets as suggested.

**Comment by MohmadAyman** [comment]
> Finish this PR. There are some general remarks in #1375 (review) to address before we can merge this into the created feature branch.

Yes I've applied that locally, ill try pushing the commit asap.

> Optionally, think about using Qt's UI files to layout widgets - instead of manually creating and layouting items.
> Fully modularize MSA widgets as suggested.

Those sounds like nice ideas 👍 

> Add loading of existing config files to allow proper editing of files. This will render the ConfigurationFilesWidget obsolete as there is no danger to overwrite manual changes.

This would be nice too from my experience implementing that into the `ros_controllers` widget, this could take some time, which I'm not sure that I do have unfortunately.

**Comment by rhaschke** [comment]
@MohmadAyman what is the status of this PR? There wasn't any activity for a while. Would be great to get this merged in soon...

**Comment by MohmadAyman** [comment]
@rhaschke can you have a look now?
I really would prefer if this gets merged into master, the PR is too big for me to manage.
Or you can continue working on if yourself if you would like, I just don't want to see a very helpful enhancement to the setup assistant(however in complete it is) go to waste.

**Comment by rhaschke** [comment]
@MohmadAyman, I will look at it later this week. I'm also very interested to get this merged :wink:

**Comment by MohmadAyman** [comment]
ping @rhaschke

**Comment by rhaschke** [comment]
It's on my todo. But, I'm extremely busy preparing weekly lectures...

**Comment by rhaschke** [comment]
I merged this into the feature branch and files an issue as a reminder for the open TODOs (#1527).

**Pull request merged**